[PM-3343] Capture TOTP QR codes from websites in the browser extension

quexten opened the pull request:
## Type of change

<!-- (mark with an `X`) -->

```
- [ ] Bug fix
- [x] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

Some sites don't provide the TOTP secret directly, but only a QR code. Additionally, copying and pasting is an additional step for the user. Finally, copying the TOTP secret temporarily exposes it to the clipboard, which might be snooped on by other applications / browser extensions. 

This PR adds a button next to the TOTP secret field, which when pressed captures a screenshot of the current webpage, extracts the TOTP QR code and reads it.

Closes https://community.bitwarden.com/t/totp-screenshot-feature/7043/2

## Code changes

- messages.json: Add the i18n messages for English
- manifest.json / manifestv3: Add <all_urls> permission for chrome.tabs.captureVisibleTab to work (on chrome, just giving "activeTab" works, but firefox requires <all_urls>)
- browser-api.ts: Add a function to capture the visible tab
- add-edit.component.ts: Add the code to capture the tab, decode the QR code and - if parsed correctly - set the cipher's TOTP secret entry
- add-edit.component.html: Add the button to capture the TOTP code from the webpage
- package.json: Add the qrcode-parser library for QR code parsing

## Screenshots
https://user-images.githubusercontent.com/11866552/258958273-463c7fc5-050a-4bff-a8ab-ee1683aeb76c.webm

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team
- Ensure that all UI additions follow [WCAG AA requirements](https://contributing.bitwarden.com/contributing/accessibility/)

quexten commented:
Right now, the PR is parsing the QR code as a URI such as:
`otpauth://totp/NodeJSTest:testuser?secret=GZ4FORKTNBVFGQTFJJGEIRDOKY&issuer=NodeJSTest`
and extracting just the secret, as on browser the default approach is copy pasting just the secret anyways. We could also store the entire URI in the TOTP secret field (this is what happens on mobile). I'm happy to change it if required, but don't want to sway this choice either way as I'm OK with both.

quexten commented:
As a note for QA: I tested manifest v2 developer builds of Firefox/Chrome. I don't have a testing setup for Safari at the moment. It could be that we need a separate way to capture the webpage screenshot on Safari, reference https://stackoverflow.com/questions/3329117/what-does-visiblecontentsasdataurl-exactly-do

sukhleenb commented:
Hello! The designers are requesting a few UI/UX tweaks to be consistent with mobile. 

Could we look into
- updating the toast copy to "Authenticator key added" when the QR code successfully scans
- updating the toast copy to "Item saved" when the user selects Save 
- on hover, add/update the camera icon tooltip to say “scan QR code”
- once the authenticator key is added, add the copy icon next to the camera icon
- double-check that when a user has "can-view" permissions, that the camera icon and the copy icon are in a disabled state

Once these changes are implemented, this will be a great win for the extension. Appreciate your work on this.

quexten commented:
> on hover, add/update the camera icon tooltip to say “scan QR code”

Thanks for the feedback. Minor nit-pick, "scan QR code" could mean that the camera is used (in case of a laptop, or a desktop with a webcam is used). How about "Scan QR code from webpage"?

Also:

> updating the toast copy to "Item saved" when the user selects Save

When saving, it does pop the notification (visible in the video). Does the point refer to replacing the active toast instead of pushing a new one?

quexten commented:
![image](https://github.com/bitwarden/clients/assets/11866552/d6d00096-8e7f-4fa1-a202-ab860e152ade)

Added the copy button; changed the messages to be:

- "Unable to scan QR code from the current webpage"
- "Authenticator key added"
- "Scan authenticator QR code from current webpage"
- "Copy TOTP secret"

> double-check that when a user has "can-view" permissions, that the camera icon and the copy icon are in a disabled state

Question about this; I can see why with "can-view" permissions, the capture would be disabled as the user should not modify the entry. But as the code is visible to the user, they could just mark & copy it, or manually copy it by copying it letter by letter. So I'm not sure what the purpose is for this?

sukhleenb commented:
- "Scan QR code from webpage" works great for the tooltip 
- For the "Item saved" toast, looks like the copy is correct. I must have misread it when viewing the video. 
- For "can-view", the main purpose is to prevent edits to the item. We are looking into disabling all quick-action buttons for this permission. If you'd like to be consistent with the existing UI, we can disable capture for now and look at disabling the copy icon with the other initiative.  
- Is the "Copy TOTP secret" the tooltip message for the copy icon? If so, can you please rename it to "Copy verification code" as seen below
<img width="479" alt="Screenshot 2023-08-08 at 2 21 49 PM" src="https://github.com/bitwarden/clients/assets/95638650/24c4acff-3875-4b4d-ab0c-6ffadbfb3a01">

quexten commented:
> Is the "Copy TOTP secret" the tooltip message for the copy icon? If so, can you please rename it to "Copy verification code" as seen below

When editing the entry, it shows the TOTP secret (`otpauth://totp/NodeJSTest:testuser?secret=GZ4FORKTNBVFGQTFJJGEIRDOKY&issuer=NodeJSTest`)
When regularly viewing the entry, it shows the generated TOTP codes (`143234`,...)

The copy button is left as "Copy TOTP code" in view mode (when actually copying TOTP codes).
In edit mode, since it shows the TOTP secret, and upon click it copies the secret, the copy button says "Copy TOTP secret" (since it is a different action from copying the generated TOTP code)

> For "can-view", the main purpose is to prevent edits to the item. We are looking into disabling all quick-action buttons for this permission. If you'd like to be consistent with the existing UI, we can disable capture for now and look at disabling the copy icon with the other initiative.

That's fine. For now I completely hid the actions (both copy the secret and capture the qr code) when in view only mode (this is the same behaviour as is currently applied for the "generate username" / "generate password buttons").

quexten commented:
Actually, thinking about it, since the text field is labeled "Authenticator key (TOTP)", it might make sense to rename the copy message to "Copy Authenticator key" or "Copy Authenticator key (TOTP)".

quexten commented:
![image](https://github.com/bitwarden/clients/assets/11866552/c7e4da62-cdbe-48cc-a707-dd6ddd7e4c0a)

Like this?

quexten commented:
In view only mode:

![image](https://github.com/bitwarden/clients/assets/11866552/74295ef3-0bbe-41f5-8b27-dee7f170fcf0)

sukhleenb commented:
I realized I misspoke. With "can view", you can currently copy all fields, regardless of if you are on the edit or view page. We should hide the copy icon for the "Can-view, except password" permission in the edit page, along with hiding the field value. We would keep the copy icon for the user to copy the verification code in the view page. Screenshots of existing UI with "Can-view, except password" permissions.

<img width="369" alt="Screenshot 2023-08-08 at 3 31 55 PM" src="https://github.com/bitwarden/clients/assets/95638650/50f548e3-ef70-46fe-89ed-5de2d57c770f">
<img width="372" alt="Screenshot 2023-08-08 at 3 31 49 PM" src="https://github.com/bitwarden/clients/assets/95638650/fb8e3697-1865-4c09-89f4-d038966c6b7b">

We'll still want to disable to capture icon for both permissions.

quexten commented:
>  We should hide the copy icon for the "Can-view, except password" permission in the edit page, along with hiding the field value

![image](https://github.com/bitwarden/clients/assets/11866552/06a6e6f8-0965-4c07-8400-d73b5821827a)

That behaviour is unchanged in this PR :)

Should the "Copy Authenticator secret (TOTP)" button be enabled when the permission is "Can-view"? (Right now both actions are disabled for both permissions).

sukhleenb commented:
> Should the "Copy Authenticator secret (TOTP)" button be enabled when the permission is "Can-view"? (Right now both actions are disabled for both permissions).

Yup, exactly! Other than that, I think this should be good to go :)

sukhleenb commented:
If possible, could you also update your video to reflect the changes discussed here?

quexten commented:
[Screencast from 2023-08-09 20-34-44.webm](https://github.com/bitwarden/clients/assets/11866552/7982e32c-b5e0-4209-bea4-354a2a538d0c)

quexten commented:
Covered most things in the video, here is a screenshot of a view only (not "except password") cipher in edit mode:
![image](https://github.com/bitwarden/clients/assets/11866552/638fa818-4049-4087-b496-98aa5659729e)

Btw, not directly related to the PR but should the authenticator key also be hidden by default like the password (since it's just as sensitive)?

sukhleenb commented:
That looks perfect, thank you! 

As for hiding the authenticator key with the "can view" permissions, it's something we can definitely look into. It is currently hidden when the permissions are set to "Can view, except passwords". That may be a larger discussion though but I'll take the feedback back to the team. 

Again, appreciate all your work on this. I'll pass it off for code review.

jonatino commented:
Any updates on the status of code review? I've been dying to have this feature on Bitwarden ever since leaving 1Password.

quexten commented on apps/browser/src/manifest.json:
```diff
@@ -52,6 +52,7 @@
     "default_popup": "popup/index.html"
   },
   "permissions": [
+    "<all_urls>",
```
I tested `activeTab` during development of the PR, however this only works on chrome/chromium-based browsers. On Firefox (and possibly Safari), `<all_urls>` is required to use the captureVisibleTab function. This is also documented here: https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/tabs/captureVisibleTab#browser_compatibility 

This seems to be a bug recorded here:
https://bugzilla.mozilla.org/show_bug.cgi?id=1784920

So the options to move forward from my point of view are:
- Staying with `<all_urls>` until the bug is fixed
- Splitting the manifest with different permissions per browser until the bug is fixed
- Not supporting the feature on Firefox until the bug is fixed

Let me know how you want to move forward with this.

Jingo88 commented on apps/web/src/app/vault/individual-vault/add-edit.component.ts:
```diff
@@ -88,6 +89,7 @@ export class AddEditComponent extends BaseAddEditComponent implements OnInit, On
       organizationService,
       sendApiService,
       dialogService,
+      window,
```
Thanks! One more build conflict. This is needed in the `super` for `apps/desktop/src/vault/app/vault/add-edit.component.ts`

quexten replied:
Whoops, tested web and browser builds but missed desktop. Should be fixed now.

andrebispo5 reviewed (approved):
Looks good!

Jingo88 commented:
Thank you for your contribution!
